containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala

```diff
@@ -0,0 +1,346 @@
+import java.nio.file.{Files, Paths}
+import java.nio.charset.Charset
+
+import org.apache.commons.io.FileUtils
+import org.apache.spark.SparkContext
+import org.apache.spark.ml.PipelineModel
+import org.json4s._
+import org.json4s.jackson.Serialization.{read, write}
+import org.json4s.jackson.Serialization
+
+import scalaj.http._
+
+import scala.sys.process._
+
+sealed trait ModelType extends Serializable
+
+case object MLlibModelType extends ModelType
+
+case object PipelineModelType extends ModelType
+
+object ModelTypeSerializer
+    extends CustomSerializer[ModelType](
+      format =>
+        (
+          {
+            case JString("MLlibModelType") => MLlibModelType
+            case JString("PipelineModelType") => PipelineModelType
+          }, {
+            case MLlibModelType => JString("MLlibModelType")
+            case PipelineModelType => JString("PipelineModelType")
+          }
+      ))
+
+class UnsupportedEnvironmentException(e: String) extends RuntimeException(e)
+
+class ModelDeploymentError(e: String) extends RuntimeException(e)
+
+case class ClipperContainerConf(var className: String,
+                                var jarName: String,
+                                var modelType: ModelType,
+                                var fromRepl: Boolean = false,
+                                var replClassDir: Option[String] = None)
+
+object Clipper {
+
+  val CLIPPER_CONF_FILENAME: String = "clipper_conf.json"
+  val CONTAINER_JAR_FILE: String = "container_source.jar"
+  val MODEL_DIRECTORY: String = "model"
+  val REPL_CLASS_DIR: String = "repl_classes"
+  // TODO: Uncomment before committing
+  //  val CLIPPER_SPARK_CONTAINER_NAME = "clipper/spark-scala-container"
+  val CLIPPER_SPARK_CONTAINER_NAME = "dcrankshaw/spark-scala-container"
+
+  val DOCKER_NW: String = "clipper_nw"
+  val CLIPPER_MANAGEMENT_PORT: Int = 1338
+
+
+  implicit val json4sFormats = Serialization.formats(NoTypeHints) + ModelTypeSerializer
+
+
+  def deploySparkModel[M](sc: SparkContext,
+                          name: String,
+                          version: Int,
+                          model: M,
+                          containerClass: Class[_],
+                          clipperHost: String,
+                          labels: List[String],
+                          sshUserName: Option[String] = None,
+                          sshKeyPath: Option[String] = None,
+                          dockerRequiresSudo: Boolean = true): Unit = {
+
+    val basePath = Paths.get("/tmp", name, version.toString).toString
+    // Use the same directory scheme of /tmp/<name>/<version> on host
+    val hostDataPath = basePath
+    val localHostNames = Set("local", "localhost", "127.0.0.1")
+    val islocalHost = localHostNames.contains(clipperHost.toLowerCase())
+
+    try {
+      saveSparkModel[M](sc, name, version, model, containerClass, basePath)
+      publishModelToClipper(clipperHost, name, version, labels, hostDataPath)
```

Comment:
Done. Moved it below the check for sudo credentials and below copying the model data to Clipper. We will only publish the model to Clipper if the ssh credentials were both supplied and have access to the machine.